### ISSUE TITLE
Consider making Encoded*Chunk.type optional

### ISSUE THREAD

**Issue by sandersdan** [w3c/webcodecs#221]
These attributes are currently required, but in some cases it can be difficult for an app to determine how to fill them. Implementations of WebCodecs are unlikely to trust or otherwise use the value, so it may be simpler to not require them.

Some signaling is going to be required for outputs from encoders, so the field should not be removed entirely (and in fact more information about dependencies may need to be added).

**Comment by padenot** [comment]
I agree, but now I wonder why it was put in in the first place, maybe I'm missing something? Maybe simply because it's useful for the output of the encoder?

**Comment by sandersdan** [comment]
`type` was always nice-to-have for decoders, but it was primarily the minimum useful information about references we could expose from encoders. The original IDL in Chrome's implementation had an optional `type`, it was changed to `required` during MSE integration work.

MSE needs this information because it fundamentally works at the level of GOPs, but it doesn't seem like MSE's requirements should be affecting the lower-level decode API.

**Comment by chcunningham** [comment]
we are now considering leaving it as required. see discussion in https://github.com/w3c/webcodecs/issues/220#issuecomment-833192313

**Comment by chcunningham** [comment]
Triage note: marking 'extension', as the proposal to go from required -> optional is not 'breaking' (even if it isn't really extending in the truest sense). 

This is purely a triage note. I still think it is unlikely we will pursue this per the previous comment.

**Comment by aboba** [comment]
A note from Jungkee Song: 

"The problem with encodedVideoChunkInit.type member of the EncodedVideoChunkInit dictionary being required is that some media transport protocols do not carry this information in the headers, so that it is necessary to parse the encoded chunk to recover it. Looking at the Chromium H.264 decoder, it’s validating the type provided by the caller, so it’s basically redoing the work to find the type in the AVC stream. Requiring this parameter seems redundant and inefficient. I'm curious if we can make the type member optional.

Looking through the history, it appears this was changed from optional to required due to some dependencies to MSE not directly related to WebCodecs. As there could be services in the wild that do not use the type info, shouldn't we consider making it optional to avoid potential perf degradation for them?"

**Comment by sandersdan** [comment]
Nit: the validation behavior is specified as SHOULD, so technically it can't be relied on by apps. The worst case is probably a decode error later, though.

**Comment by chcunningham** [comment]
> Nit: the validation behavior is specified as SHOULD, so technically it can't be relied on by apps. The worst case is probably a decode error later, though.

Oh, totally agree. I just mean they can rely on zero-validation happening outside of the SHOULD case. That is, they only need to be be certain they have a key frame at the required points.

**Comment by jungkees** [comment]
@chcunningham, @sandersdan, thanks for the reasoning of how you landed the decision and the pointer to [the other issue thread](https://github.com/w3c/webcodecs/issues/220#issuecomment-833192313).

While I understand the points, I noticed some developers have no feasible backwards compatible way to add metadata such as a frame type to communicate attributes of encoded video frames on the wire. This means they'd need to have this ability on the client side.

Can you please expand the idea that you mentioned as #4 in [Clarify flush() behavior · Issue #220 · w3c/webcodecs (github.com)](https://github.com/w3c/webcodecs/issues/220#issuecomment-833192313)? Do you think that's a feasible and performant option you can consider?

Or can we think about providing a platform API (hypothetically _isKeyFrame(data)_) so developers can access the key frame type info from the platform directly?

**Comment by sandersdan** [comment]
> I noticed some developers have no feasible backwards compatible way to add metadata such as a frame type to communicate attributes of encoded video frames on the wire. This means they'd need to have this ability on the client side.

I'd be interested to discuss specific situations where this is a problem, most wire formats do include this information.

> Can you please expand the idea that you mentioned as https://github.com/w3c/webcodecs/pull/4 in https://github.com/w3c/webcodecs/issues/220#issuecomment-833192313? Do you think that's a feasible and performant option you can consider?

The option to create or recommend a JS library isn't better defined than what is written there, just that if developers are going to to it client-side it would be nice to share the code for doing so.

Within a specific application's requirements it's usually easy to detect keyframes. For example with H.264 Annex B content it's likely sufficient to scan for the fixed byte sequence representing a NALU header for an IDR slice.

> Or can we think about providing a platform API (hypothetically isKeyFrame(data)) so developers can access the key frame type info from the platform directly?

This is trickier, as it exposes all the cases where it's not so easy. The most obvious one is encrypted content, which may be unknown or ambiguous, but also there are more complex cases like non-IDR SEI recovery points, or content with supplementary streams.

It's more likely that we would extend WebCodecs with some sort of non-keyframe handling algorithm, if this is a big enough problem that it is harming adoption of WebCodecs.
